File: execution/cua/task/yaml_task.py
from __future__ import annotations

import re
from pathlib import Path
from typing import Any

import yaml

from cua.models.task import TaskManifest

PLACEHOLDER_PATTERN = re.compile(r"\{\{([a-z][a-z0-9-]*)\}\}")
RECORDED_TASK_NAME_PATTERN = re.compile(
    r"^(step-(\d{3,})) \| (click|doubleClick|input|keyboard|wait)$"
)
RECORDED_INPUT_ID_PATTERN = re.compile(r"^(step-(\d{3,}))-input$")


def read_yaml_document(path: Path) -> dict[str, Any]:
    try:
        value = yaml.safe_load(path.read_text(encoding="utf-8"))
    except Exception as error:
        raise ValueError(f"读取并解析 Midscene YAML 失败：{path}\n{error}") from error
    if not isinstance(value, dict):
        raise ValueError(f"Midscene YAML 根节点必须是对象：{path}")
    validate_yaml_document(value, path)
    return value


def validate_yaml_document(document: dict[str, Any], source: Path | str) -> None:
    tasks = document.get("tasks")
    if not isinstance(tasks, list) or not tasks:
        raise ValueError(f"Midscene YAML tasks 必须是非空数组：{source}")
    for index, task in enumerate(tasks, start=1):
        if not isinstance(task, dict):
            raise ValueError(f"Midscene YAML tasks[{index}] 必须是对象：{source}")
        name = task.get("name")
        flow = task.get("flow")
        if not isinstance(name, str) or not name.strip():
            raise ValueError(f"Midscene YAML tasks[{index}].name 不能为空：{source}")
        if not isinstance(flow, list) or not flow:
            raise ValueError(f"Midscene YAML tasks[{index}].flow 必须是非空数组：{source}")
        for action_index, action in enumerate(flow, start=1):
            if not isinstance(action, dict) or not action:
                raise ValueError(
                    f"Midscene YAML tasks[{index}].flow[{action_index}] 必须是非空对象：{source}"
                )


def dump_yaml_document(document: dict[str, Any]) -> str:
    return yaml.safe_dump(
        document,
        allow_unicode=True,
        sort_keys=False,
        default_flow_style=False,
        width=120,
    )


def validate_recorded_task_document(
    document: dict[str, Any],
    manifest: TaskManifest,
    source: Path | str,
) -> None:
    agent = document.get("agent")
    if not isinstance(agent, dict):
        raise ValueError(f"录制任务 YAML agent 必须是对象：{source}")
    if agent.get("groupName") != manifest.task:
        raise ValueError(f"录制任务 YAML agent.groupName 必须等于 task.json 的 task：{source}")
    if agent.get("groupDescription") != manifest.goal:
        raise ValueError(f"录制任务 YAML agent.groupDescription 必须等于 task.json 的 goal：{source}")

    previous_step_number = 0
    operation_by_step_id: dict[str, str] = {}
    for index, task in enumerate(document["tasks"], start=1):
        name = task["name"]
        match = RECORDED_TASK_NAME_PATTERN.fullmatch(name)
        if match is None:
            raise ValueError(
                f"录制任务 YAML tasks[{index}].name 必须符合 step-NNN | <operation-type>：{source}"
            )
        step_id, number_text, operation_type = match.groups()
        step_number = int(number_text)
        if step_number <= 0 or step_id != f"step-{step_number:03d}":
            raise ValueError(f"录制任务 YAML tasks[{index}] 的 step ID 非法：{source}")
        if step_number <= previous_step_number:
            raise ValueError(f"录制任务 YAML step ID 必须唯一且严格递增：{source}")
        if task.get("continueOnError") is True:
            raise ValueError(f"录制任务 YAML 不允许启用 continueOnError：{source}")
        operation_by_step_id[step_id] = operation_type
        previous_step_number = step_number

    for input_id in manifest.inputs:
        match = RECORDED_INPUT_ID_PATTERN.fullmatch(input_id)
        if match is None:
            raise ValueError(f"录制任务输入 ID 必须符合 step-NNN-input：{input_id}")
        step_id, number_text = match.groups()
        step_number = int(number_text)
        if step_number <= 0 or step_id != f"step-{step_number:03d}":
            raise ValueError(f"录制任务输入 ID 非法：{input_id}")
        if operation_by_step_id.get(step_id) != "input":
            raise ValueError(f"录制任务输入 {input_id} 未对应 input 类型步骤")


def write_yaml_document(path: Path, document: dict[str, Any]) -> None:
    validate_yaml_document(document, path)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(dump_yaml_document(document), encoding="utf-8")


def collect_placeholders(value: object) -> set[str]:
    if isinstance(value, str):
        matches = set(PLACEHOLDER_PATTERN.findall(value))
        residue = PLACEHOLDER_PATTERN.sub("", value)
        if "{{" in residue or "}}" in residue:
            raise ValueError(f"非法输入占位符：{value}")
        return matches
    if isinstance(value, list):
        return set().union(*(collect_placeholders(item) for item in value)) if value else set()
    if isinstance(value, dict):
        placeholders: set[str] = set()
        for key, item in value.items():
            placeholders.update(collect_placeholders(key))
            placeholders.update(collect_placeholders(item))
        return placeholders
    return set()


def resolve_yaml_inputs(
    document: dict[str, Any],
    manifest: TaskManifest,
    provided_inputs: dict[str, str] | None = None,
) -> tuple[dict[str, Any], dict[str, str]]:
    provided = provided_inputs or {}
    unknown_inputs = sorted(set(provided) - set(manifest.inputs))
    if unknown_inputs:
        raise ValueError(f"未知输入参数：{', '.join(unknown_inputs)}")

    placeholders = collect_placeholders(document)
    undeclared = sorted(placeholders - set(manifest.inputs))
    if undeclared:
        raise ValueError(f"YAML 包含未声明输入占位符：{', '.join(undeclared)}")
    unused = sorted(set(manifest.inputs) - placeholders)
    if unused:
        raise ValueError(f"任务清单输入未在 YAML 中使用：{', '.join(unused)}")

    values = {input_id: definition.default for input_id, definition in manifest.inputs.items()}
    values.update(provided)

    def resolve(value: object) -> object:
        if isinstance(value, str):
            return PLACEHOLDER_PATTERN.sub(lambda match: values[match.group(1)], value)
        if isinstance(value, list):
            return [resolve(item) for item in value]
        if isinstance(value, dict):
            return {resolve(key): resolve(item) for key, item in value.items()}
        return value

    resolved = resolve(document)
    assert isinstance(resolved, dict)
    validate_yaml_document(resolved, "resolved task")
    if collect_placeholders(resolved):
        raise ValueError("resolved task 仍包含未解析输入占位符")
    return resolved, values

File: execution/cua/task/yaml-task.ts
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import path from 'node:path';
import { parseYamlScript } from '@midscene/core/yaml';
import { parse, stringify } from 'yaml';
import type { JsonObject, TaskManifest } from '../contracts/types.js';

const placeholderPattern = /\{\{([a-z][a-z0-9-]*)\}\}/g;
const recordedTaskNamePattern = /^(step-(\d{3,})) \| (click|doubleClick|input|keyboard|wait)$/;
const recordedInputIdPattern = /^(step-(\d{3,}))-input$/;

function isRecord(value: unknown): value is JsonObject {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value);
}

export function validateYamlDocument(document: JsonObject, source: string): void {
  const tasks = document.tasks;
  if (!Array.isArray(tasks) || tasks.length === 0) {
    throw new Error(`Midscene YAML tasks 必须是非空数组：${source}`);
  }
  tasks.forEach((task, taskIndex) => {
    if (!isRecord(task)) throw new Error(`Midscene YAML tasks[${taskIndex + 1}] 必须是对象：${source}`);
    if (typeof task.name !== 'string' || !task.name.trim()) {
      throw new Error(`Midscene YAML tasks[${taskIndex + 1}].name 不能为空：${source}`);
    }
    if (!Array.isArray(task.flow) || task.flow.length === 0) {
      throw new Error(`Midscene YAML tasks[${taskIndex + 1}].flow 必须是非空数组：${source}`);
    }
    task.flow.forEach((action, actionIndex) => {
      if (!isRecord(action) || Object.keys(action).length === 0) {
        throw new Error(
          `Midscene YAML tasks[${taskIndex + 1}].flow[${actionIndex + 1}] 必须是非空对象：${source}`,
        );
      }
    });
  });
}

export async function readYamlDocument(sourcePath: string): Promise<JsonObject> {
  const absolutePath = path.resolve(sourcePath);
  let document: unknown;
  try {
    document = parse(await readFile(absolutePath, 'utf8'));
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    throw new Error(`读取并解析 Midscene YAML 失败：${absolutePath}\n${message}`);
  }
  if (!isRecord(document)) throw new Error(`Midscene YAML 根节点必须是对象：${absolutePath}`);
  validateYamlDocument(document, absolutePath);
  return document;
}

export function dumpYamlDocument(document: JsonObject): string {
  validateYamlDocument(document, 'YAML output');
  return stringify(document, { lineWidth: 120 });
}

export async function writeYamlDocument(sourcePath: string, document: JsonObject): Promise<void> {
  const absolutePath = path.resolve(sourcePath);
  await mkdir(path.dirname(absolutePath), { recursive: true });
  await writeFile(absolutePath, dumpYamlDocument(document), 'utf8');
}

export function validateRecordedTaskDocument(
  document: JsonObject,
  manifest: TaskManifest,
  source: string,
): void {
  if (!isRecord(document.agent)) throw new Error(`录制任务 YAML agent 必须是对象：${source}`);
  if (document.agent.groupName !== manifest.task) {
    throw new Error(`录制任务 YAML agent.groupName 必须等于 task.json 的 task：${source}`);
  }
  if (document.agent.groupDescription !== manifest.goal) {
    throw new Error(`录制任务 YAML agent.groupDescription 必须等于 task.json 的 goal：${source}`);
  }

  let previous = 0;
  const operationByStep = new Map<string, string>();
  (document.tasks as JsonObject[]).forEach((task, index) => {
    const match = recordedTaskNamePattern.exec(String(task.name));
    if (!match) {
      throw new Error(`录制任务 YAML tasks[${index + 1}].name 必须符合 step-NNN | <operation-type>：${source}`);
    }
    const [, stepId, numberText, operationType] = match;
    const number = Number(numberText);
    if (number <= 0 || stepId !== `step-${String(number).padStart(3, '0')}`) {
      throw new Error(`录制任务 YAML tasks[${index + 1}] 的 step ID 非法：${source}`);
    }
    if (number <= previous) throw new Error(`录制任务 YAML step ID 必须唯一且严格递增：${source}`);
    if (number !== index + 1) throw new Error(`录制任务 YAML step ID 必须从 step-001 开始连续编号：${source}`);
    if (task.continueOnError === true) throw new Error(`录制任务 YAML 不允许启用 continueOnError：${source}`);
    operationByStep.set(stepId, operationType);
    previous = number;
  });

  for (const inputId of Object.keys(manifest.inputs)) {
    const match = recordedInputIdPattern.exec(inputId);
    if (!match) throw new Error(`录制任务输入 ID 必须符合 step-NNN-input：${inputId}`);
    const [, stepId, numberText] = match;
    const number = Number(numberText);
    if (number <= 0 || stepId !== `step-${String(number).padStart(3, '0')}`) {
      throw new Error(`录制任务输入 ID 非法：${inputId}`);
    }
    if (operationByStep.get(stepId) !== 'input') {
      throw new Error(`录制任务输入 ${inputId} 未对应 input 类型步骤`);
    }
  }

  if (manifest.source.stepBindings) {
    const taskStepIds = [...operationByStep.keys()];
    const bindingStepIds = Object.keys(manifest.source.stepBindings);
    const missing = taskStepIds.filter((stepId) => !Object.hasOwn(manifest.source.stepBindings!, stepId));
    const extra = bindingStepIds.filter((stepId) => !operationByStep.has(stepId));
    if (missing.length || extra.length) {
      throw new Error(`录制证据绑定必须与 YAML 步骤一致：missing=${missing.join(',') || '-'} extra=${extra.join(',') || '-'}`);
    }
    const recorded = Object.values(manifest.source.stepBindings).filter((value): value is number => value !== null);
    if (new Set(recorded).size !== recorded.length) throw new Error('录制证据绑定中的 trace step 不得重复');
  }
}

function collectPlaceholders(value: unknown): Set<string> {
  if (typeof value === 'string') {
    const matches = new Set([...value.matchAll(placeholderPattern)].map((match) => match[1]));
    const residue = value.replace(placeholderPattern, '');
    if (residue.includes('{{') || residue.includes('}}')) throw new Error(`非法输入占位符：${value}`);
    return matches;
  }
  if (Array.isArray(value)) {
    return new Set(value.flatMap((item) => [...collectPlaceholders(item)]));
  }
  if (isRecord(value)) {
    const result = new Set<string>();
    for (const [key, item] of Object.entries(value)) {
      for (const match of collectPlaceholders(key)) result.add(match);
      for (const match of collectPlaceholders(item)) result.add(match);
    }
    return result;
  }
  return new Set();
}

export function resolveYamlInputs(
  document: JsonObject,
  manifest: TaskManifest,
  provided: Record<string, string> = {},
): { document: JsonObject; inputs: Record<string, string> } {
  const unknown = Object.keys(provided).filter((key) => !Object.hasOwn(manifest.inputs, key)).sort();
  if (unknown.length) throw new Error(`未知输入参数：${unknown.join(', ')}`);
  const placeholders = collectPlaceholders(document);
  const undeclared = [...placeholders].filter((key) => !Object.hasOwn(manifest.inputs, key)).sort();
  if (undeclared.length) throw new Error(`YAML 包含未声明输入占位符：${undeclared.join(', ')}`);
  const unused = Object.keys(manifest.inputs).filter((key) => !placeholders.has(key)).sort();
  if (unused.length) throw new Error(`任务清单输入未在 YAML 中使用：${unused.join(', ')}`);

  const inputs = Object.fromEntries(
    Object.entries(manifest.inputs).map(([inputId, definition]) => [inputId, definition.default]),
  );
  Object.assign(inputs, provided);
  const resolve = (value: unknown): unknown => {
    if (typeof value === 'string') {
      return value.replace(placeholderPattern, (_, inputId: string) => inputs[inputId]);
    }
    if (Array.isArray(value)) return value.map(resolve);
    if (isRecord(value)) return Object.fromEntries(Object.entries(value).map(([key, item]) => [resolve(key), resolve(item)]));
    return value;
  };
  const resolved = resolve(document);
  if (!isRecord(resolved)) throw new Error('resolved task 根节点不是对象');
  validateYamlDocument(resolved, 'resolved task');
  if (collectPlaceholders(resolved).size) throw new Error('resolved task 仍包含未解析输入占位符');
  parseYamlScript(dumpYamlDocument(resolved));
  return { document: resolved, inputs };
}
